src/components/atoms/NumberInput/BaseNumberInput.tsx

```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+
+import { NumberInputProps } from './NumberInput';
+
+import { Container, Input } from './styles';
+
+
+const BaseNumberInput = React.forwardRef<HTMLInputElement, NumberInputProps>((props, ref) => {
+  const {
+    SuffixComponent = null,
+    PrefixComponent = null,
+    value = 0,
+    onChange,
+    min = Number.NEGATIVE_INFINITY,
+    max = Number.POSITIVE_INFINITY,
+    step = 1,
+    allowSpecialCharacters = false,
+    disableDecimal = false,
+    onKeyDown = () => {},
+    variant = 'default'
+  } = props;
+
+  const numberValue = Number(value);
+
+
+  const _onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    onChange(e);
+  };
+
+
+  const _onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+
+    if ((allowSpecialCharacters && [ '+', '-', 'e' ].includes(e.key)) || (disableDecimal && [ '.' ].includes(e.key))) {
+      e.preventDefault();
+    }
+
+
+    //detect down arrow key
+    if (e.key === 'ArrowDown') {
+      onDecrement();
+    }
+
+    if (e.key === 'ArrowUp') {
+      onIncrement();
+    }
+
+    onKeyDown(e);
+  };
+
+
+  const onIncrement = () => {
+    if (max >= numberValue + step) {
+      const increasedVal = numberValue + step; //this one increases above next value, so floor it
+      const floorValue = Math.floor(increasedVal / step) * step;
+
+      // @ts-ignore
```

Comment:
why this comment

Reply:
since onChange is controlled we have to generate an event object